Sources_v3/Controllers/UserController/UserController_Tests.swift

```diff
@@ -393,6 +393,160 @@ final class UserController_Tests: StressTestCase {
         // Assert updater is called with correct `userId`
         XCTAssertEqual(env.userUpdater!.unmuteUser_userId, controller.userId)
     }
+    
+    // MARK: - Flag user
+    
+    func test_flagUser_propogatesError() {
+        // Simulate `flag` call and catch the completion.
+        var completionError: Error?
+        controller.flag { [callbackQueueID] in
+            AssertTestQueue(withId: callbackQueueID)
+            completionError = $0
+        }
+        
+        // Simulate network response with the error.
+        let networkError = TestError()
+        env.userUpdater!.flagUser_completion!(networkError)
+        
+        // Assert error is propogated.
+        AssertAsync.willBeEqual(completionError as? TestError, networkError)
+    }
+    
+    func test_flagUser_propogatesNilError() {
+        // Simulate `flag` call and catch the completion.
+        var completionIsCalled = false
+        controller.flag { [callbackQueueID] error in
+            // Assert callback queue is correct.
+            AssertTestQueue(withId: callbackQueueID)
+            // Assert there is no error.
+            XCTAssertNil(error)
+            completionIsCalled = true
+        }
+        
+        // Simulate successful network response.
+        env.userUpdater!.flagUser_completion!(nil)
+        
+        // Assert completion is called.
+        AssertAsync.willBeTrue(completionIsCalled)
+    }
+    
+    func test_flagUser_callsUserUpdater_withCorrectValues() {
+        // Simulate `flag` call.
+        controller.flag()
+        
+        // Assert updater is called with correct `flag`
+        XCTAssertEqual(env.userUpdater!.flagUser_flag, true)
+        // Assert updater is called with correct `userId`
+        XCTAssertEqual(env.userUpdater!.flagUser_userId, controller.userId)
+    }
+    
+    func test_flagUser_keepsControllerAlive() {
+        // Simulate `flag` call and catch the completion.
+        var completionIsCalled = false
+        controller.flag { [callbackQueueID] error in
+            // Assert callback queue is correct.
+            AssertTestQueue(withId: callbackQueueID)
+            // Assert there is no error.
+            XCTAssertNil(error)
+            completionIsCalled = true
+        }
+        
+        // Create a weak ref and release a controller.
+        weak var weakController = controller
+        controller = nil
+        
+        // Assert controller is kept alive
+        AssertAsync.staysTrue(weakController != nil)
+        
+        // Simulate successful network response.
+        env.userUpdater!.flagUser_completion!(nil)
+        
+        // Assert completion is called.
+        AssertAsync.willBeTrue(completionIsCalled)
+        
+        // Clean up the mock updater that keeps a callback closure with the controller.
+        env.userUpdater!.cleanUp()
```

Comment:
I see what you're doing here but I find it a bit confusing. What about doing it in a more straightforward way:
```swift
        // Simulate a successful network response.
         env.userUpdater!.flagUser_completion!(nil)
         // The recorded completion block has a strong reference to controller, we need to reset it:
         env.userUpdater!.flagUser_completion = nil
```

A little bit nicer solution would be to introduce API like:
```swift
extension UserUpdaterMock {
    /// Simulates API response for all recorded completion blocks and resets them.
    func simulateAPICallResponse(_ error: Error?) {
        muteUser_completion?(error)
        muteUser_completion = nil
        
        unmuteUser_completion?(error)
        unmuteUser_completion = nil
        
        // etc ...
    }
}
```

Reply:
Regarding the snippet you shared, I'm not sure about it to be honest: 
1. We don't reset other values related to some method -> there will be a mix of nil and non-nil values.
2. When we do `env.userUpdater!.flagUser_completion!(nil)` we get a crash if the completion we are invoking is nil. It's useful to quickly understand that something is wrong.
3. This method doesn't work if we want to call some completion > 1 time if we performed 2 API calls.
4. we still need `env.userUpdater.cleanUp` in `tearDown` because there are some tests where we check that updater is called with correct values but we don't simulate a response (`test_flagUser_callsUserUpdater_withCorrectValues/test_unflagUser_callsUserUpdater_withCorrectValues`)

Reply:
My main point here is that we keep the completion block just to call it. Once we call it, it's no longer needed. This is IMO implementation detail of the mock and we shouldn't expose it. Ideally, the mock has the same behavior as the real implementation - and the _real_ updater doesn't keep the closure alive after the call is finished.

I hear your points about my snippet and they make sense. Happy to see your alternatives to the issue I mentioned above.

Reply:
If you think it's not worth spending time on time feel free to keep it. The PR is approved :)

Reply:
Yep, I suggest to go with the `cleanUp` we already familiar with. If we decide we are not OK with it, then we can have a ticket and fix it thought all the tests at once